File: 98-validate-binary-search-tree/in-order.js
let prev, curr

/**
 * 中序遍历，检查前后结点是否严格有序
 * 时间：O(N), 68ms
 * 空间：O(H), 37.4MB
 */
var isValidBST = function(root) {
  prev = curr = null
  return inOrder(root)
};

function inOrder (node) {
  if (!node) return true

  if (!inOrder(node.left)) return false
  
  prev = curr
  curr = node
  if (prev && curr && prev.val >= curr.val) return false

  if (!inOrder(node.right)) return false

  return true
}

File: problems/897-increasing-order-search-tree/in-order.cpp
/**
 * 重新串联
 * 
 * 时间：`O(N)`, 28ms
 */

class Solution {
public:
    TreeNode* increasingBST(TreeNode* root) {
        TreeNode* dummy = new TreeNode(-1);
        chain = dummy;
        inOrder(root);
        TreeNode* res = dummy->right;
        dummy->right = NULL;
        delete dummy;
        return res;
    }

    void inOrder(TreeNode* node) {
        if (!node) return;

        TreeNode* left = node->left;
        TreeNode* right = node->right;
        // 左
        inOrder(left);
        // 串联本结点
        node->left = node->right = NULL;
        chain->right = node;
        chain = chain->right;
        // 右
        inOrder(right);
    }

private:
    TreeNode* chain;
};